modules/collect/src/main/java/com/opengamma/strata/collect/array/ObjArray.java

```diff
@@ -0,0 +1,929 @@
+/**
+ * Copyright (C) 2016 - present by OpenGamma Inc. and the OpenGamma group of companies
+ * 
+ * Please see distribution for license.
+ */
+package com.opengamma.strata.collect.array;
+
+import java.io.Serializable;
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Type;
+import java.util.AbstractList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.ListIterator;
+import java.util.Map;
+import java.util.NoSuchElementException;
+import java.util.RandomAccess;
+import java.util.Set;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+import java.util.function.IntFunction;
+import java.util.stream.Stream;
+
+import org.joda.beans.Bean;
+import org.joda.beans.BeanBuilder;
+import org.joda.beans.ImmutableBean;
+import org.joda.beans.JodaBeanUtils;
+import org.joda.beans.MetaBean;
+import org.joda.beans.MetaProperty;
+import org.joda.beans.Property;
+import org.joda.beans.PropertyStyle;
+import org.joda.beans.impl.BasicImmutableBeanBuilder;
+import org.joda.beans.impl.BasicMetaBean;
+import org.joda.beans.impl.BasicMetaProperty;
+
+import com.google.common.base.Objects;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.opengamma.strata.collect.ArgChecker;
+import com.opengamma.strata.collect.function.IntObjConsumer;
+import com.opengamma.strata.collect.function.IntObjFunction;
+
+/**
+ * An immutable array of {@code Object} values.
+ * <p>
+ * This provides a cut down version of {@link List} that implements {@link Array}.
+ * <p>
+ * It is intended that the array does not contain {@code null}, however this is not validated.
+ * 
+ * @param <T>  the type of the element
+ */
+public final class ObjArray<T>
```

Comment:
Given my comment about keeping `ScenarioResult` as the return type of calculation functions, I don't think we need this class at all. `DefaultScenarioResult` does as much as we need. Although maybe it could do with a better name.